feat: Custom host location

Udit-takkar opened the pull request:
## What does this PR do?

Fixes: https://github.com/calcom/cal.com/issues/25709

1) Assign users to team event type and enable this toggle

https://github.com/user-attachments/assets/f03760cc-6858-4425-af05-7f3ea849bc90


2) Select location app for each user and test various scenarios


## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.









<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Adds per-host custom locations for round-robin event types so each host can set their own meeting location. Bookings use the organizer’s location, auto-link matching credentials, and fall back to Cal Video when needed.

- **New Features**
  - Toggle in Event Setup to enable per-host locations for round-robin.
  - Host-by-host location picker with mass-apply, organizer input dialogs, and app availability badges.
  - Disables the global Locations picker and hides the public location list when per-host locations are enabled.
  - Booking service uses the organizer host’s location, auto-links a matching credential, and falls back to Cal Video when needed.
  - TRPC endpoints to load hosts with current location and installed apps, and to mass-apply a location.
  - Persists locations via a new HostLocation model and the enablePerHostLocations flag.

- **Migration**
  - Adds EventType.enablePerHostLocations and creates HostLocation table and indexes.
  - Run database migrations; no manual steps required.

<sup>Written for commit de814d82d31d143dddcd1f322b93e2758205f739. Summary will update on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

Udit-takkar committed:
chore: save progress

Udit-takkar committed:
Merge branch 'main' into feat/custom-location

Udit-takkar committed:
chore:

CarinaWolli commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -4163,5 +4163,11 @@
+  "host_locations": "Host Locations",
+  "apply_to_all_hosts": "Apply to all hosts",
+  "select_location": "Select location",
+  "host_locations_fallback_description": "When a host doesn't have the selected app installed, Cal Video will be used as a fallback",
```
I think it would be better if the fallback would be the location that is set above, here: 
<img width="486" height="203" alt="Screenshot 2026-01-08 at 1 41 12 PM" src="https://github.com/user-attachments/assets/57bf1091-43e2-4bc2-a152-60c564a40b42" />

CarinaWolli reviewed (commented):
Early review

Udit-takkar committed:
feat: add input dialog

devin-ai-integration[bot] committed:
Merge branch 'main' into feat/custom-location

Resolved merge conflicts:
- apps/web/modules/event-types/components/tabs/setup/EventSetupTab.tsx: Updated imports to use new paths
- packages/trpc/server/routers/viewer/eventTypes/types.ts: Merged hostLocationSchema and enablePerHostLocations

Additional type fixes:
- Added location field to host fallback objects in roundRobinReassignment.ts and roundRobinManualReassignment.ts
- Updated getHostsWithLocationOptions.handler.ts to use credentialForCalendarServiceSelect
- Added enablePerHostLocations to defaultEvents.ts commons object

Co-Authored-By: unknown <>

Udit-takkar committed:
Merge branch 'main' into feat/custom-location

Udit-takkar committed:
fix: type error

Udit-takkar committed:
Merge branch 'main' into feat/custom-location

Udit-takkar committed:
feat: mass apply dialog

Udit-takkar committed:
test: per host location

Udit-takkar committed:
test: fix test

Udit-takkar committed:
Merge branch 'main' into feat/custom-location

devin-ai-integration[bot] committed:
fix: address Cubic AI review feedback (confidence >= 9/10)

- Remove PII (address, phone number) from tracing logs in RegularBookingService.ts
- Constrain HostLocation.type to EventLocationType["type"] for compile-time validation

Co-Authored-By: unknown <>

Udit-takkar committed:
fix: translation

joeauyeung commented on apps/web/modules/event-types/components/locations/HostLocations.tsx:
```diff
@@ -0,0 +1,664 @@
+  locationOptions: TLocationOptions;
+};
+
+const getAppSlugFromLocationType = (locationType: string): string | null => {
```
Should these helper functions belong to `/app-store/locations`? For example, there's already a static options enum in there so it would act as a single source of truth.

Udit-takkar replied:
Ok. I can move it there

joeauyeung commented on apps/web/modules/event-types/components/locations/HostLocations.tsx:
```diff
@@ -0,0 +1,664 @@
+  const enablePerHostLocations = formMethods.watch("enablePerHostLocations");
+  const hosts = formMethods.watch("hosts");
+
+  const { data: hostsWithApps, isLoading } = trpc.viewer.eventTypes.getHostsWithLocationOptions.useQuery(
```
Curious on your thoughts. For scalability, should this call not be batched with the initial trpc calls? I'm already seeing slowdowns on event types with a large number of hosts and if this is batched in the initial call trpc call it'll slow down the initial load time for the event type form.

joeauyeung commented on packages/features/bookings/lib/service/RegularBookingService.ts:
```diff
@@ -1299,9 +1301,105 @@ async function handler(
+  let perHostCredentialId: number | undefined = undefined;
+
+  // Handle per-host custom locations for round-robin events
+  if (
```
I think this should live in a separate helper function. I'm not a fan of adding large chunks of logic like this to the main handler.

Udit-takkar replied:
I'll move it to a class

joeauyeung reviewed (commented):
Awesome work on this @Udit-takkar. I do have some feedback.

# Blocking
- Curious if you thought about scalability with this feature enabled and multiple hosts. I want to make sure that this doesn't increase the load times of the event type form if this option is enabled.
- With the custom host locations enabled we still show the location option. If we're falling back to Cal Video anyways I think we should hide the location option as it might cause some confusion
<img width="563" height="535" alt="Image" src="https://github.com/user-attachments/assets/4994c1f6-4978-4072-b378-59fbe980d209" />
- When all hosts locations are set to one that requires attendee input, we're not getting that from the attendee on the booking page. I think we should just disable those types of options for this feature

# Non-blocking
- On the booking page, it defaults to "Cal Video" as the location even if it's not a selection

Udit-takkar committed:
Merge branch 'main' into feat/custom-location

Udit-takkar committed:
refactor: improvements

devin-ai-integration[bot] committed:
fix: correct grammar in custom host locations tooltip

Change 'custom host locations is enabled' to 'custom host locations are enabled' (plural subject requires plural verb).

Addresses Cubic AI review feedback (confidence 9/10).

Co-Authored-By: unknown <>

Udit-takkar committed:
refactor: improvements

Udit-takkar committed:
fix: auth

Udit-takkar committed:
fix: check

sean-brydon reviewed (dismissed):
<img width="2096" height="360" alt="Image" src="https://github.com/user-attachments/assets/3ff2fa87-94e9-4fcd-afed-bd9e12eff457" />

Hey Udit - this works super well super great feature nice work! 

Looking at the design side of this i think theres a few things we can improve. 

* https://github.com/calcom/cal.com/blob/f11d6edf7df932bd16c72d9ef6dc921a97abbc67/apps/web/modules/event-types/components/locations/HostLocations.tsx#L647

I wonder if it makes sense to figure out a way to still have this load more feature but not show the h-1 gap as it creates this weird space below the list items. 

* <img width="694" height="196" alt="Image" src="https://github.com/user-attachments/assets/97a4768e-307b-41c5-9438-6f999cfaf064" />

This label "configure locations per host" has like 46pixels of margin bottom applied we looks super weird 

*  <img width="716" height="628" alt="Image" src="https://github.com/user-attachments/assets/37a458ca-4352-4910-beae-902d3db98145" />

Not too sure whats happened with avatar here

sean-brydon commented on packages/trpc/server/routers/viewer/eventTypes/getHostsWithLocationOptions.handler.ts:
```diff
@@ -0,0 +1,141 @@
+  ctx,
+  input,
+}: GetHostsWithLocationOptionsInput): Promise<GetHostsWithLocationOptionsResponse> => {
+  const { eventTypeId, cursor, limit } = input;
```
Not sure if im being very stupid here but it looks like this has no auth attached. Any authed user can just pass in an eventTypeId

sean-brydon commented on packages/trpc/server/routers/viewer/eventTypes/_router.ts:
```diff
@@ -145,4 +147,26 @@ export const eventTypesRouter = router({
     });
   }),
+
+  getHostsWithLocationOptions: authedProcedure
```
Should this also use createEventPBAC?

sean-brydon commented on packages/features/bookings/lib/service/RegularBookingService.ts:
```diff
@@ -19,6 +19,7 @@ import {
   OrganizerDefaultConferencingAppType,
 } from "@calcom/app-store/locations";
 import { getAppFromSlug } from "@calcom/app-store/utils";
+import { HostLocationRepository } from "@calcom/features/host/repositories/HostLocationRepository";
```
Unused

Udit-takkar committed:
refactor: improvements

devin-ai-integration[bot] committed:
fix: address Cubic AI review feedback (confidence >= 9/10)

- Add scheduleId to newly created hosts in update.handler.ts to persist
  host-specific schedules during create operations
- Change host location deletion filter from !host.location to
  host.location === null to only delete when explicitly set to null
- Fix static-link per-host locations to use actual link instead of type
  in locationBodyString for bookingLocationService.ts

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
fix: preserve existing host scheduleId when not explicitly provided

Change scheduleId handling for existing hosts from 'host.scheduleId ?? null'
to 'host.scheduleId === undefined ? undefined : host.scheduleId' so that
when the client doesn't provide a scheduleId, the existing value is preserved
instead of being cleared to null.

Co-Authored-By: unknown <>

Udit-takkar committed:
fix; type erro

Udit-takkar committed:
fix; type erro

Udit-takkar committed:
fix; type erro

joeauyeung reviewed (dismissed):
Thanks for addressing my previous feedback @Udit-takkar. I just see prisma queries in `packages/trpc/server/routers/viewer/eventTypes/massApplyHostLocation.handler.ts`. Let's move these to a repository before approving.

Udit-takkar committed:
refactor: move repository

Udit-takkar committed:
refactor: move repository

devin-ai-integration[bot] committed:
fix: add singular/plural translations for location_applied_to_hosts

Addresses Cubic AI review feedback (confidence 9/10) to fix '1 hosts' rendering as '1 host' by using i18next plural format with _one and _other suffixes.

Co-Authored-By: unknown <>

Udit-takkar committed:
refactor: feedback

Udit-takkar committed:
fix: type err

hariombalhara commented:
@Udit-takkar I think we should remove the locations that require Booker Input from the custom host location dropdown for now. This is because those locations require Booking Questions to be updated so that we can show an input field there. It doesn't work right now with custom host locations

hariombalhara commented on packages/features/bookings/lib/service/RegularBookingService.ts:
```diff
@@ -1299,9 +1300,40 @@ async function handler(
+      organizerOrFirstDynamicGroupMemberDefaultLocationUrl = result.organizerDefaultLocationUrl;
+      perHostCredentialId = result.perHostCredentialId;
+
+      tracingLogger.info("Using per-host location", {
```
This could be a lot of logs in production, this should probably be debug

hariombalhara reviewed (dismissed):
There is just one blocking feedback. We should use UUID data type for id column

Udit-takkar committed:
Merge branch 'main' into feat/custom-location

Udit-takkar committed:
fix: use uuid in schema and remove attendee locaiton

Udit-takkar committed:
fix: type err

Udit-takkar committed:
fix: type err

devin-ai-integration[bot] committed:
fix: validate eventTypeId as integer in massApplyHostLocation schema

Co-Authored-By: unknown <>

Udit-takkar commented on apps/web/modules/event-types/components/locations/HostLocations.tsx:
```diff
@@ -0,0 +1,841 @@
+              userId: host.userId,
+              eventTypeId,
+              type: locationType,
+              credentialId: null,
```
@cubic-dev-ai query is re fetched and form is then updated with correct credential

devin-ai-integration[bot] committed:
Merge branch 'main' into feat/custom-location

Co-Authored-By: udit@cal.com <udit222001@gmail.com>

joeauyeung reviewed (dismissed):
All feedback is addressed. Great work @Udit-takkar
